src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/TransferReadTags.java

```diff
@@ -0,0 +1,154 @@
+ * base and adds the read tags from the unaligned bam to the aligned bam.
+ *
+ * Currently, the tool is implemented for the specific case of transfering UMI read tags (RX)
+ * from an unaligned bam.
+ */
+@CommandLineProgramProperties(
+        summary = "Incorporate read tags in a SAM file to that of a matching SAM file",
+        oneLineSummary = "Incorporate read tags in a SAM file to that of a matching SAM file",
+        programGroup = QCProgramGroup.class // sato: change
+)
+public class TransferReadTags extends GATKTool {
+    final static String UMI_TAG_NAME = "RX";
+
+    @Argument(fullName = "unmapped-sam", doc = "query-name sorted unmapped sam file containing the read tag of interest")
+    public GATKPath unmappedSamFile;
+    ReadsDataSource unmappedSam;
+
+    @Argument(shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, doc = "")
+    public File outSamFile;
+    SAMFileGATKReadWriter writer;
+
+    @Argument(fullName = "read-tags", doc = "read tag names to transfer")
+    public List<String> readTags = new ArrayList<>();
+
+
+    PeekableIterator<GATKRead> alignedSamIterator;
+    PeekableIterator<GATKRead> unmappedSamIterator;
+
+    GATKRead currentRead1;
+    GATKRead currentRead2;
+    ReadQueryNameComparator queryNameComparator;
+
+    public void onTraversalStart(){
+        Utils.nonEmpty(readTags, "read tags may not be empty");
+        queryNameComparator = new ReadQueryNameComparator();
+        alignedSamIterator = new PeekableIterator<>(directlyAccessEngineReadsDataSource().iterator());
+        SAMFileHeader.SortOrder sortOrder1 = directlyAccessEngineReadsDataSource().getHeader().getSortOrder();
+        Utils.validate(sortOrder1 == SAMFileHeader.SortOrder.queryname, "aligned sam must be sorted by queryname");
+
+        unmappedSam = new ReadsPathDataSource(unmappedSamFile.toPath());
+
+        // We would check that the unmapped bam is query-name sorted, but
+        // the SortOrder field is often not populated. Thus we simply assume that
+        // the unmapped sam is query-name sorted and rely on the traversal to throw an
+        // error when it is not.
+        unmappedSamIterator = new PeekableIterator<>(unmappedSam.iterator());
+
+        // Initialize Read2. Read1 will be initialized in traverse()
+        if (unmappedSamIterator.hasNext()){
+            currentRead2 = unmappedSamIterator.next();
+        } else {
+            throw new UserException("unaligned sam iterator is empty.");
+        }
+
+        writer = createSAMWriter(new GATKPath(outSamFile.getAbsolutePath()), false);
+    }
+
+    @Override
+    public void traverse() {
+        // will need to use this later
+        while (alignedSamIterator.hasNext()){
+            currentRead1 = alignedSamIterator.next();
+            int diff = queryNameComparator.compareReadNames(currentRead1, currentRead2);
```

Comment:
Let me ask you. Is there a reason you cant just pull on both iterators in tandem and blow up if there is ever a mismatch between the reads and samflags? Do you expect the unmapped bam to have more reads in it for some reason? I might consider reading `AbstractAlignmentMerger.mergeAlignment()` in picard to see how they handle iterating the two lists in tandem.

Reply:
Excellent point. STAR aligner drops some reads when it creates the transcriptome-aligned bam, so I cannot assume that there are an equal number of reads in both aligned and unaligned. I need to think more carefully about the case you brought up though.

Reply:
OK phew it look like samtools query-name sorts the way you describe, and Picard SortSam (which is what we are using in the RNA pipeline) does the string compare (see `SAMRecordQueryNameComparator`). So the code as is should work; how about we add the samtools version of comparator after this week's release (since we don't need it for this pipeline if not ever)

Reply:
This might or might not be an issue? we don't support writing bams labeled as queryname ordered in the samtools sense but we do support reading them? That will mean you get cryptic errors here if you aren't careful. Its also somewhat of an edge cases that likely won't show up in your pipelines